### PR TITLE
fix(rustc_lint): better detect when parens are necessary

### LINKED ISSUE #90807: `unused_parens` false positive for parenthesised `Range` in `for` loop
Given the following code: [playground stable](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=70bddfd331c49d1a8868a751584994ea) / [nightly](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c7fd6a85bcc24c5ceee440e0b7afdeb9)
```rust
fn main() {
    for _ in (1..{ 2 }) {}
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
warning: unnecessary parentheses around `for` iterator expression
 --> src/main.rs:2:14
  |
2 |     for _ in (1..{ 2 }) {}
  |              ^        ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
2 -     for _ in (1..{ 2 }) {}
2 +     for _ in 1..{ 2 } {}
  | 

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 1.91s
     Running `target/debug/playground`
```

No warnings should be generated as removing the parentheses around the `Range` expression changes the meaning of the code (to a compiler error):

```rust
fn main() {
    for _ in 1..{ 2 } {}
}
```
The suggested code is parsed as a `RangeFrom` and `{ 2 }` as the `for` loop’s block, followed by en empty block:
```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:2:19
  |
2 |     for _ in 1..{ 2 } {}
  |                   ^ expected `()`, found integer

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` due to previous error
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e316ad6ae538c8ea9fc049a11bd2accb)

---

## `rustc` versions
* stable 1.56.1
* nightly (installed via rustup)
```console
$ rustc --version --verbose
rustc 1.58.0-nightly (8b09ba6a5 2021-11-09)
binary: rustc
commit-hash: 8b09ba6a5d5c644fe0f1c27c7f9c80b334241707
commit-date: 2021-11-09
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by notriddle**
Fixes #90807

### CODE REVIEW

File: src/test/ui/lint/unused/issue-90807-unused-paren.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// check-pass
+// Make sure unused parens lint doesn't emit a false positive.
+// See https://github.com/rust-lang/rust/issues/90807
+#![deny(unused_parens)]
+
+fn main() {
+    for _ in (1..{ 2 }) {}
```
Comment (nagisa) [inline src/test/ui/lint/unused/issue-90807-unused-paren.rs:7]: Could you also add tests for other kinds of combinations? For example:

```rust
for _ in 1..loop { break 2 } {}
```

```rust
for _ in 1..match () { () => 2 } {}
```

both work and should continue warning, but I think after this PR they won't.

Comment (notriddle) [reply src/test/ui/lint/unused/issue-90807-unused-paren.rs:7]: Okay, I've tweaked it to do the right thing with the loop and match expressions.

**Comment by nagisa** [comment]
@bors r+

Thanks!

**Pull request merged**